OAK-10162: Fix Version copier with preserveOnTarget to ignore diverge…

amit-jain opened the pull request:
…d history

When preserveOnTarget is true
- Ignoring diverged version history or if no change detected
- Not changing version properties if these properties are null in the source which prevents linkage to new successor versions being broken

Commit:
OAK-10162: Fix Version copier with preserveOnTarget to ignore diverged history

When preserveOnTarget is true
- Ignoring diverged version history or if no change detected
- Not changing version properties if these properties are null in the source which prevents linkage to new successor versions being broken

mreutegg reviewed (commented):
Looks good. I have some minor suggestions for improvements: https://github.com/apache/jackrabbit-oak/pull/884

rishabhdaim commented on oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/migration/NodeStateCopier.java:
```diff
@@ -145,29 +150,73 @@ public static boolean copyNodeStore(@NotNull final NodeStore source, @NotNull fi
         return hasChanges;
     }
+    
+    private static Set<String> getValues(NodeState nodeState, String prop) {
+        if (nodeState.hasProperty(prop)) {
+            Iterable<String> values = nodeState.getProperty(prop).getValue(Type.STRINGS);
+            return StreamSupport.stream(values.spliterator(), false).filter(s -> !s.isEmpty()).collect(Collectors.toSet());
+        }
+        return new HashSet<>();
```
Implementation of this method only uses the size of `set` returned by this method. So instead of creating an `set` of non empty properties, we should simply return a boolean indicating whether it is empty or not.
```suggestion
    private static boolean isVersionEmpty(NodeState nodeState, String prop) {
        if (nodeState.hasProperty(prop)) {
            Iterable<String> values = nodeState.getProperty(prop).getValue(Type.STRINGS);
            return StreamSupport.stream(values.spliterator(), false).allMatch(String::isEmpty);
        }
        return true;
```

amit-jain replied:
I had left it to leave room for furture in case needed but not a strong opinion here.

rishabhdaim commented on oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/migration/NodeStateCopier.java:
```diff
@@ -145,29 +150,73 @@ public static boolean copyNodeStore(@NotNull final NodeStore source, @NotNull fi
+            if (property.getName().equals(JcrConstants.JCR_UUID) || 
+                property.getName().equals(JcrConstants.JCR_SUCCESSORS) || 
+                property.getName().equals(JcrConstants.JCR_PREDECESSORS)) {
+                boolean versionPropertyEmpty = getValues(source, property.getName()).isEmpty();
```
```suggestion
                boolean versionPropertyEmpty = isVersionEmpty(source, property.getName());
```

rishabhdaim commented on oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/migration/NodeStateCopier.java:
```diff
@@ -145,29 +150,73 @@ public static boolean copyNodeStore(@NotNull final NodeStore source, @NotNull fi
+    
+    private static boolean isVersionPropertyEmpty(NodeState source, PropertyState property, boolean preserveOnTarget,
+        String path) {
+        if (preserveOnTarget) {
```
I would simply return if `preserveOnTarget` is false.
```suggestion
        if (!preserveOnTarget) {
        return false;
```
then to go inside if condition.
As a general coding practice, I found:

`if not condition:
  do little stuff
  return
// now...
do lots of stuff.`

increases the code readability and reduce cyclomatic complexity.

amit-jain replied:
Here it is not much stuff to do and also adds another if condition but again not a strong opinion.

rishabhdaim commented on oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/migration/version/VersionCopier.java:
```diff
@@ -111,6 +125,72 @@ public boolean copyVersionHistory(String versionableUuid, Calendar minDate, bool
 
+    private boolean hasNoConflicts(String versionHistoryPath, String versionableUuid, boolean preserveOnTarget, NodeState sourceVersionHistory) {
+        // if preserveOnTarget is true then check no conflicts which means version history has moved forward only
+        if (preserveOnTarget) {
```
same as above, return if `preserveOnTarget` is false.

amit-jain replied:
same reasoning as above, the code is currently concise and change will end up having multiple if blocks

rishabhdaim commented on oak-core/src/test/java/org/apache/jackrabbit/oak/plugins/migration/version/VersionCopierTest.java:
```diff
@@ -44,18 +50,15 @@
-import static org.apache.jackrabbit.JcrConstants.JCR_UUID;
-import static org.apache.jackrabbit.JcrConstants.MIX_VERSIONABLE;
-import static org.apache.jackrabbit.JcrConstants.NT_UNSTRUCTURED;
+import static org.apache.jackrabbit.JcrConstants.*;
```
Please import them individually.

rishabhdaim commented on oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/migration/version/VersionCopier.java:
```diff
@@ -111,6 +125,72 @@ public boolean copyVersionHistory(String versionableUuid, Calendar minDate, bool
+                VersionComparator versionComparator = new VersionComparator();
+
+                // version history id not equal
+                boolean conflictingVersionHistory = !targetVersionHistory.getString(JCR_UUID).equals(sourceVersionHistory.getString(JCR_UUID));
```
Since `NodeBuilder.getString()` can return null, I would compare them using `Objects.equal`

amit-jain replied:
we are in the version history node path which will mandatorily contain jcr:uuid property but Objects.equal comparison is good.

rishabhdaim commented on oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/migration/version/VersionCopier.java:
```diff
@@ -111,6 +125,72 @@ public boolean copyVersionHistory(String versionableUuid, Calendar minDate, bool
+
+                // highest source version UUID does not match the corresponding version on target (diverged)
+                boolean conflictingHighestVersion =
+                    !sourceVersionHistory.getChildNode(sourceVersions.get(0)).getString(JCR_UUID).equals(targetVersionHistory.getChildNode(sourceVersions.get(0)).getString(JCR_UUID));
```
Please use `Objects.equal`

rishabhdaim commented on oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/migration/version/VersionCopier.java:
```diff
@@ -111,6 +125,72 @@ public boolean copyVersionHistory(String versionableUuid, Calendar minDate, bool
+        return true;
+    }
+
+    static class VersionComparator implements Comparator<String> {
```
We are making a reverse comparison here. 
Shouldn't we make a simple comparison and then where we are using we should explicitly call `comparator.reversed()` to highlight that we are using a reverse comparison, wdyt?

amit-jain replied:
Maybe adding javadoc will convey the same purpose

Commit:
OAK-10162: Fix Version copier with preserveOnTarget to ignore diverged history

- Incorporating changes suggested in review

mreutegg committed:
OAK-10162: Fix Version copier with preserveOnTarget to ignore diverge… (#884)

Do not create new HashSet for empty values
Use getProperty() and then check for null instead of has/getProperty()
Simplify test by using existing code to remove a version
